testutil/dbustest2.go

```diff
@@ -0,0 +1,232 @@
+
+	m        sync.Mutex
+	readable sync.Cond
+
+	outputBuf, inputBuf bytes.Buffer
+	closed              bool
+	authDone            bool
+	n                   int
+}
+
+func (s *testDBusStream) decodeRequest() {
+	// s.m is locked
+	if !s.authDone {
+		// Before authentication is done process the text protocol anticipating
+		// the TEST authentication used by NewDBusTestConn call below.
+		msg := s.inputBuf.String()
+		s.inputBuf.Reset()
+		switch msg {
+		case "\x00":
+			// initial NUL byte, ignore
+		case "AUTH\r\n":
+			s.outputBuf.WriteString("REJECTED TEST\r\n")
+		case "AUTH TEST TEST\r\n":
+			s.outputBuf.WriteString("OK test://\r\n")
+		case "CANCEL\r\n":
+			s.outputBuf.WriteString("REJECTED\r\n")
+		case "BEGIN\r\n":
+			s.authDone = true
+		default:
+			panic(fmt.Errorf("unrecognized authentication message %q", msg))
+		}
+		s.readable.Signal()
+		return
+	}
+
+	// After authentication the buffer must contain marshaled DBus messages.
+	msgIn, err := dbus.DecodeMessage(&s.inputBuf)
+	if err != nil {
+		panic(fmt.Errorf("cannot decode incoming message: %v", err))
+	}
+	switch s.n {
+	case 0:
+		// The very first message we receive is a Hello message sent from
+		// NewDBusTestConn below. This message, along with the NameAcquired
+		// signal we send below, allow DBus to associate the client with the
+		// responses sent by test.
+		s.sendMsg(&dbus.Message{
+			Type: dbus.TypeMethodReply,
+			Headers: map[dbus.HeaderField]dbus.Variant{
+				dbus.FieldDestination: dbus.MakeVariant(testDBusClientName),
+				dbus.FieldSender:      dbus.MakeVariant("org.freedesktop.DBus"),
+				dbus.FieldReplySerial: dbus.MakeVariant(msgIn.Serial()),
+				dbus.FieldSignature:   dbus.MakeVariant(dbus.SignatureOf("")),
+			},
+			Body: []interface{}{":test"},
+		})
+		s.sendMsg(&dbus.Message{
+			Type: dbus.TypeSignal,
+			Headers: map[dbus.HeaderField]dbus.Variant{
+				dbus.FieldPath:        dbus.MakeVariant(dbus.ObjectPath("/org/freedesktop/DBus")),
+				dbus.FieldInterface:   dbus.MakeVariant("org.freedesktop.DBus"),
+				dbus.FieldMember:      dbus.MakeVariant("NameAcquired"),
+				dbus.FieldDestination: dbus.MakeVariant(testDBusClientName),
+				dbus.FieldSender:      dbus.MakeVariant("org.freedesktop.DBus"),
+				dbus.FieldSignature:   dbus.MakeVariant(dbus.SignatureOf("")),
+			},
+			Body: []interface{}{testDBusClientName},
+		})
+	default:
+		msgOutList, err := s.handler(msgIn, s.n-1)
+		if err != nil {
+			panic("cannot handle message: " + err.Error())
+		}
+		for _, msgOut := range msgOutList {
+			// Test code does not need to provide the address of the sender.
+			if _, ok := msgOut.Headers[dbus.FieldSender]; !ok {
+				msgOut.Headers[dbus.FieldSender] = dbus.MakeVariant(testDBusClientName)
+			}
+			s.sendMsg(msgOut)
+		}
+	}
+	s.n++
+}
+
+func (s *testDBusStream) sendMsg(msg *dbus.Message) {
+	// TODO: handle big endian if we ever get big endian machines again.
+	if err := msg.EncodeTo(&s.outputBuf, binary.LittleEndian); err != nil {
```

Comment:
godbus seems to use a native endian https://github.com/godbus/dbus/blob/a81f0b52152ffcc4c163b8f13a2609b989e23028/transport_unix.go#L196 though it should be transparent even if we use little endian here as long as we call `DecodeMessage()` in the receive part.

Reply:
If only go exported native endianness without forcing everyone to copy-paste the pointer trick. ...

Reply:
I actually like https://github.com/yalue/native_endian, it's nice because it has a compile-time constant answer.